### PR TITLE
feat(confluence): add Confluence integration plugin (scaffold)

### LINKED ISSUE #282: feat(plugin): Add Confluence integration
### Description
Confluence is a widely used corporate knowledge base. Agents resolving customer issues or writing internal documentation need to query wiki spaces, read pages, and update documentation. Adding a Confluence plugin ensures agents have access to internal company wikis securely.

### Proposed Package Structure
A new package under `packages/confluence`:
- `packages/confluence/endpoints/`:
  - `pages.ts`: Actions to retrieve page content (`pages.get`), search across spaces (`pages.search`), or create/update page content (`pages.create`, `pages.update`).
  - `spaces.ts`: List or fetch metadata about spaces (`spaces.list`).

### Use Cases
- "Find the onboarding documentation in Confluence and summarize the setup steps."
- "Create a new page under the Q3 goals space with the following meeting notes."

### PR DESCRIPTION
**Pull request by rishabh1244**
## Description

Fixes #282.

Adds the initial Confluence Cloud integration under `packages/confluence`. This scaffold provides three read endpoints and production-ready API-token and OAuth 2.0 authentication, leaving write operations and webhooks for focused follow-up PRs.

### What's included

- `pages.get` - list/retrieve pages through the Confluence v2 API
- `pages.search` - search content across spaces using CQL
- `spaces.list` - list Confluence spaces with filtering
- API-token Basic auth using the required Atlassian account email
- Atlassian OAuth 2.0 (3LO) with Cloud-ID gateway routing, rotating refresh tokens, and explicit multi-site selection
- Zod input/output schemas on every endpoint
- Pagination parameters (`limit`, `cursor`/`start`) on list endpoints
- Rate-limit handling with exponential backoff

### Not included (planned for follow-up PRs)

The full target lists 62 operations and 23 webhook triggers. This PR intentionally ships only three read endpoints and their authentication foundation. Write operations, attachments, comments, labels, and webhook triggers will follow separately. Confluence Cloud webhook support is not represented by a placeholder here because production Connect webhooks require their own installation and JWT verification flow.

### Validation

- Endpoint invocation tests covering every implemented endpoint with a mocked client
- Authentication tests covering OAuth metadata, refresh-token rotation, Cloud-ID site selection, and Basic credential construction
- Client tests covering Atlassian gateway routing and multi-site behavior
- Schema validation tests
- Live integration tests are opt-in and skipped when credentials are absent

## Screenshots / Demos

Test suite passing locally:

Confluence plugin test suite passing

## Checklist

- [x] Plugin scaffolded with `pnpm generate:plugin`
- [x] Provider registered in `packages/corsair/core/constants.ts`
- [x] Input/output Zod schemas on all endpoints
- [x] Pagination params on list endpoints
- [x] Tests with assertions (no live-only tests in CI)
- [x] No boilerplate residue (TODOs, commented-out code, placeholder URLs)
- [x] Local lint, typecheck, build, tests, and plugin validation pass

### CODE REVIEW

File: packages/confluence/endpoints/pages.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+export const get: ConfluenceEndpoints['pagesGet'] = async (ctx, input) => {
+	const validated = PagesGetInputSchema.parse(input);
+
+	const cloudUrl =
+		ctx.options.cloudUrl ?? (await ctx.keys.get_cloud_url()) ?? '';
+
+	const result = await makeConfluenceRequest<
+		ConfluenceEndpointOutputs['pagesGet']
+	>('../../api/v2/pages', ctx.key, cloudUrl, {
```
Comment (ambikeesshh) [inline packages/confluence/endpoints/pages.ts]: This `'../../api/v2/pages'` path is the concern in point 2 of the top-level summary. `request()` in `corsair/http` concatenates `${baseUrl}/${path}`, so the resolved URL ends up as `${cloudUrl}/wiki/rest/api/../../api/v2/pages`. Verify it works against real Confluence Cloud; consider overriding BASE for v2 endpoints.

File: packages/confluence/webhooks/types.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		return parsedBody !== null && parsedBody.type === eventType;
+	};
+}
+
+export function verifyConfluenceWebhookSignature(
+	request: WebhookRequest<ConfluenceWebhookPayload>,
+	secret: string,
+): { valid: boolean; error?: string } {
+	// TODO: Implement webhook signature verification
+	return { valid: true };
```
Comment (ambikeesshh) [inline packages/confluence/webhooks/types.ts]: Hardcoded `{ valid: true }`. Any payload gets accepted as authentic. Point 3 in the top-level summary. Either implement real signature verification or drop the webhook subsystem from this PR.

File: pnpm-lock.yaml [hunk 3]
```diff
@@ -614,7 +614,34 @@ importers:
...
+        specifier: ^8.0.1
+        version: 8.5.1(jiti@2.7.0)(postcss@8.5.15)(tsx@4.22.4)(typescript@5.9.3)(yaml@2.9.0)
+      typescript:
+        specifier: 'catalog:'
+        version: 5.9.3
+      zod:
+        specifier: 4.4.3
+        version: 4.4.3
+
+  packages/confluence:
```
Comment (ambikeesshh) [inline pnpm-lock.yaml:652]: CI install is failing here: `ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY` for `ts-jest@29.4.9(...)`. Point 1 in the top-level summary. Regenerate via `pnpm install` and commit the result.

**Review by ambikeesshh — CHANGES_REQUESTED** [review]
Three blockers before this can merge:

1. `pnpm-lock.yaml` is broken: CI fails at install with `ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY` for `ts-jest@29.4.9(...)`, so lint/typecheck/build/test never actually ran. The lockfile diff also has ~25 unrelated `ts-jest` line edits across other importers. Re-run `pnpm install` on the branch and commit the regenerated file.

2. The `pages.get` URL path (`'../../api/v2/pages'`) is fragile. `corsair/http`'s `request()` does plain string concat, so the final URL is `${cloudUrl}/wiki/rest/api/../../api/v2/pages`. The sibling endpoints (`space`, `search`) sit under the v1 tree as plain relative paths, but this one needs to escape up two levels to reach `/wiki/api/v2/pages`. Worth confirming it actually resolves against a real Confluence Cloud instance in the demo. If it does, a cleaner pattern would be overriding BASE per-call for v2 endpoints, or having the client accept an absolute path.

3. `verifyConfluenceWebhookSignature` in `webhooks/types.ts` always returns `{ valid: true }`. That means every incoming payload is treated as authentic regardless of signature. Either implement real verification (the exact scheme depends on whether this is Atlassian Connect, native webhooks, or Forge), or drop the webhook subsystem from this PR and land it in a follow-up.

File: packages/confluence/schema/database.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { z } from 'zod';
```
Comment (ambikeesshh) [inline packages/confluence/schema/database.ts]: Whole file is commented-out boilerplate from the generator. R5 wants this gone. Either delete the file or define the real entities here.

File: packages/confluence/index.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+		webhookHooks: options.webhookHooks,
+		endpoints: confluenceEndpointsNested,
+		webhooks: confluenceWebhooksNested,
+		endpointMeta: confluenceEndpointMeta,
+		endpointSchemas: confluenceEndpointSchemas,
+		webhookSchemas: confluenceWebhookSchemas,
+		pluginWebhookMatcher: (request) => {
+			const headers = request.headers;
+			// TODO: Update to match your webhook signature headers
+			return 'x-confluence-signature' in headers;
```
Comment (ambikeesshh) [inline packages/confluence/index.ts]: `x-confluence-signature` isn't a real Confluence webhook header, and there's a `// TODO` right above it. This matcher is effectively a placeholder. Needs the real header name for whatever webhook mechanism this integration targets (Atlassian Connect, native webhooks, or Forge).

**Review by ambikeesshh — COMMENTED** [review]
A few boilerplate leftovers that should get cleaned up before merge (PR rule R5: no generator residue):

- `schema/database.ts` is entirely commented-out scaffolding. Either define the real entities there or delete the file.
- The webhook matcher in `index.ts` checks for a header named `x-confluence-signature`, which isn't a real Confluence webhook header. There's also a `// TODO` left in the code. Needs real research into how Atlassian signs payloads for whatever webhook type this integration uses (Atlassian Connect, native webhooks, or Forge).
- Minor: the `pages.search` description in the endpoint meta is a copy-paste from `pages.get` (both say "List Confluence pages"). Should probably be something like "Search Confluence pages via CQL".

File: packages/confluence/index.ts [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+const defaultAuthType: AuthTypes = 'api_key' as const;
+
+const confluenceEndpointMeta = {
+	'pages.get': {
+		riskLevel: 'read',
+		description: 'List Confluence pages',
+	},
+	'pages.search': {
+		riskLevel: 'read',
+		description: 'List Confluence pages',
```
Comment (ambikeesshh) [inline packages/confluence/index.ts]: Description here looks copy-pasted from `pages.get` above. `pages.search` should probably say something like "Search Confluence pages via CQL".

File: packages/confluence/api.test.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import 'dotenv/config';
+import { Pages, Spaces } from './endpoints';
+import { ConfluenceEndpointOutputSchemas } from './endpoints/types';
+import type { ConfluenceContext } from './index';
+
+const API_KEY = process.env.CONFLUENCE_API_KEY!;
```
Comment (ambikeesshh) [inline packages/confluence/api.test.ts]: Reading env vars with `!` and no guard means missing creds produce a confusing 401 deep in the request layer. A `beforeAll` that skips the suite with a clear message when `CONFLUENCE_API_KEY` / `CONFLUENCE_CLOUD_URL` are missing would be friendlier.

File: packages/confluence/integration.test.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+	});
+
+	it('pages.get returns correct type', async () => {
+		const result = await Pages.get(mockCtx, { limit: 10 });
+		ConfluenceEndpointOutputSchemas.pagesGet.parse(result);
+	});
+
+	it('pages.search returns correct type', async () => {
+		const result = await Pages.search(mockCtx, {
+			cql: 'type = "page" AND space = "SD"',
```
Comment (ambikeesshh) [inline packages/confluence/integration.test.ts:35]: Only `{ limit: 10 }` is passed here, so the `space_id` / `title` / `status` / `cursor` filtering paths in `pages.get` aren't exercised. Worth adding a case or two.

**Review by ambikeesshh — COMMENTED** [review]
Some smaller notes, mostly on the test side. Not blocking.

- `api.test.ts` reads `process.env.CONFLUENCE_API_KEY!` and `CONFLUENCE_CLOUD_URL!` with the non-null assertion but never guards them. If the env vars are missing, the value is `undefined`, the request proceeds with `key: undefined`, and you get a confusing 401 deep inside the HTTP layer. A `beforeAll` that skips the suite with a clear message when those vars are absent would be friendlier.
- The test for `pages.get` only passes `{ limit: 10 }`, so the filtering paths (`space_id`, `title`, `status`, `cursor`) aren't exercised. Worth adding a case or two if you can.
- R2 mentions slack's `api.test.ts` + `integration.test.ts` combo as an example. Right now this PR has `api.test.ts` (real API) and `schema.test.ts` (two trivial assertions). Consider adding at least one mocked test so CI can validate logic without hitting the real API.

**Comment by rishabh1244** [comment]
Thanks for pointing out , I've made the mentioned changes

## Summary of changes

Here's everything addressed since the last review round:

### 1. Lockfile regenerated (aff21c73)
- Ran `pnpm install` to fix `ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY` for `ts-jest@29.4.9`
- Includes `packages/confluence` entry + expected ts-jest resolution updates (esbuild peer dep from tsup)

### 2. Webhook signature verification (`webhooks/types.ts`)
- Replaced `// TODO: Implement… return { valid: true }` with real HMAC-SHA256 verification
- Reads `x-hub-signature` header (Atlassian standard, same as Jira)
- Uses `crypto.timingSafeEqual` for constant-time comparison
- Gracefully allows unsigned requests when no secret is configured

### 3. API base path — removed `../../` traversal hack (`client.ts`, `endpoints/pages.ts`)
- Added optional `base` parameter to `makeConfluenceRequest`
- Default stays `/wiki/rest/api` (v1) — `spaces.ts`, `search.ts` unchanged
- `pages.ts` now uses `base: '/wiki/api/v2'` with path `'pages'` instead of `'../../api/v2/pages'`
- Resolves to the same URL without relying on path traversal

### 4. API test env guard (`api.test.ts`)
- Removed `!` non-null assertions, added `beforeAll` that throws a clear message when `CONFLUENCE_API_KEY` / `CONFLUENCE_CLOUD_URL` are missing

### 5. Webhook unit tests (`webhooks.test.ts`)
- 12 tests covering: valid signature, wrong signature, missing header, array header, length mismatch, empty secret, rawBody priority, payload fallback, and `createConfluenceMatch` edge cases

### 6. Plugin webhook matcher (`index.ts`)
- Changed `x-confluence-signature` (fictional) → `x-atlassian-webhook-identifier` (real Atlassian webhook header, same as Jira)

### 7. Pages.get filter test (`api.test.ts`)
- Added `status: 'current'` test case to exercise the filtering path
- ### 8. Removed `schema/database.ts`
- File was entirely commented-out boilerplate from `pnpm generate:plugin`
- As noted by the reviewer ("R5 wants this gone"), keeping dead code serves no purpose

**Review by ambikeesshh — COMMENTED** [review]
Reviewed and pushed fixes on top of the latest head

- Fixed broken lockfile (missing `ts-jest` peer dependency); regenerated so CI installs pass.
- Removed accidentally committed `pnpm-original-lock.yaml` (26k-line backup).
- R1: webhook signature verification now rejects requests when no secret is configured; now uses `verifyHmacSignatureWithPrefix`
- Cleaned up webhook tenant-link + matcher files (removed TODOs/commented code)
- Added `operations.test.ts` covering all 3 endpoint invocations (mocked client, no live secrets).
- Minor: fixed double-space in `pages.search` description

all CI checks are green, and greptile is clean(5/5)

**Review by ambikeesshh — COMMENTED** [review]
LGTM now. ready to merge i believe @devjain32

**Comment by ambikeesshh** [comment]
@greptileai review

File: packages/confluence/endpoints/spaces.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+		ctx.options.authType === 'oauth_2'
+			? ((await ctx.keys.get_cloud_id()) ?? undefined)
+			: undefined;
+
+	const result = await makeConfluenceRequest('space', ctx.key, cloudUrl, {
+		method: 'GET',
+		authType: ctx.options.authType,
+		cloudId,
+		query: {
+			...(validated.key && { spaceKey: validated.key }),
```
[the remark this answers is not in the document]
Comment (ambikeesshh) [reply packages/confluence/endpoints/spaces.ts]: Fixed in ebb0f1cf by moving spaces.list to the current v2 /wiki/api/v2/spaces route and using its documented keys query parameter. This also avoids relying on the deprecated v1 filter contract.

File: packages/confluence/index.ts [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+				if (result.refreshed) {
+					await Promise.all([
+						ctx.keys.set_access_token(result.accessToken),
+						ctx.keys.set_expires_at(String(result.expiresAt)),
+						ctx.keys.set_refresh_token(result.refreshToken),
+					]);
+				}
+
+				const configuredCloudUrl = options.cloudUrl ?? storedCloudUrl;
+				const normalizedConfiguredUrl = configuredCloudUrl?.replace(/\/+$/, '');
```
[the remark this answers is not in the document]
Comment (ambikeesshh) [reply packages/confluence/index.ts]: Fixed in ebb0f1cf. The key builder now reuses the linear-time URL normalization helper, removing the second polynomial-regex path.

File: packages/confluence/client.ts [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+	return {
+		accessToken: tokenData.access_token,
+		refreshToken: tokenData.refresh_token ?? refreshToken,
+		expiresAt: now + tokenData.expires_in,
+		refreshed: true,
+	};
+}
+
+function normalizeCloudUrl(cloudUrl: string): string {
+	return cloudUrl.trim().replace(/\/+$/, '');
```
[the remark this answers is not in the document]
Comment (ambikeesshh) [reply packages/confluence/client.ts]: Fixed in ebb0f1cf. Trailing slash removal now uses a single linear scan instead of a quantified regular expression, with a regression test covering long slash suffixes.

**Review by devjain32 — APPROVED** [review]
